doc/pins/p4orch_hld.md

```diff
@@ -0,0 +1,235 @@
+* Get number of entries in the map for an object type
+
+### KeyGenerator
+
+The KeyGenerator is a common module to generate keys for all managers which are used in the managers’ local map and in the centralized mapper. The following are some example key formats:
+
+* RouterInterfaceManager: router_interface_id=&lt;id>
+* NeighborManager: router_interface_id=&lt;id>,neighbor_id=&lt;id>
+* NextHopManager: next_hop_id=&lt;id>
+* RouteManager: vrf_id=&lt;id>,ipv4_dst=&lt;id> or vrf_id=&lt;id>,ipv6_dst=&lt;id>
+
+### P4Orch Interaction With Existing Orchagents
+
+When a P4Orch manager needs information about the existing SONiC objects, it will interact with the existing orchs. Most existing orchs provide public methods and global pointers for other orchs to access the SONiC objects that they manage. The P4Orch managers interact with other orchs in the following manner:
+
+* Query SONiC objects information, such as OID, existence etc.
+* Increase the reference counting of a SONiC object when it is referenced in a P4RT object.
+* Decrease the reference counting of a SONiC object when it is dereferenced from a P4RT object.
+* Increase the resource counting in CrmOrch when a SAI object is created.
+* Decrease the resource counting in CrmOrch when a SAI object is deleted.
+
+In the initial release, P4Orch managers will interact with the following SONiC orchs:
+
+* PortsOrch \
+PortsOrch provides port information and port reference counting. P4Orch managers that require port information will interact with PortsOrch.
+* VrfOrch \
+VrfOrch provides VRF information and reference counting. Currently, only RouterInterfaceManager interacts with VrfOrch.
+* CrmOrch \
+CrmOrch provides resource management counting information. Several managers interact with CrmOrch for resource counting reference when doing SAI object creation and deletion.
+
+Currently, public methods provided by P4Orch are limited. Existing SONiC orchs do not have full access to the P4RT objects.
+
+### P4Orch To Application Response Path
+
+As P4RT interface requires the response of each programming request to match the real hardware state, synchronous mode between orchagent and syncd are enabled so P4Orch can get the exact SAI API response. And P4Orch implements a notification channel to send the response further up to the P4RT application. For more details, please refer to the [APPL STATE DB & Response Path HLD](appl_state_db_response_path_hld.md) document.
+
+## SAI API
+
+N/A
+
+## Configuration and management
+
+N/A
+
+## Warmboot and Fastboot Design Impact
+
+Warmboot and fastboot support in P4Orch are being planned for the next phase (next SONiC release). In this phase, P4Orch does not support warmboot and fastboot features. There will be no impact on warmboot and fastboot of the existing orchs.
+
+## Restrictions/Limitations
+
+The P4Orch is designed to meet the SDN requirements. And hence there are some differences from the other SONiC orchagents:
```

Comment:
Also specify how the re-programming happens if lets say there is an swss crash/restart and APP_DB is cleared during init.  Call out if there are any limitations.

Reply:
Thanks.

Updated the "Restrictions/Limitations" section that P4Orch does not support warmboot and orchagent restart in the initial phase.

For my understanding, orchagent restart is similar to warmboot. In both case, the APPL DB tables won't be cleared. And orchagent will re-program the APPL DB tables in "init view" mode. This is not specific to P4Orch. It is a common feature for all orchs. Our main challenge for supporting warmboot now is that P4Orch doesn't do retry. During warmboot init, all table requests will come in a single batch. Since P4Orch doesn't do retry, dependency on SONiC table (such as port & vrf) can not be satisfied. We do not support warmboot in the initial phase.

Reply:
orchagent restart is *not* similar to warmboot. APP_DB shall be flushed during init. In regular cases, the configs are reprogrammed by managers. In case of p4rt, does this leave the system in such a state? What actions to reprogram the device?

Reply:
Sorry about the late reply. (I didn't use my primary email to setup my github account. Need to do a better job in that.)

In our use case, we don't expect the system to recover after orchagent crashes. The system will go to "critical state". The controller will be notified and starts to drain and reboot the switch. Currently, we don't have special handling in recovering P4Orch after crashing.

I would like to understand more on how SONiC handles the crashing. This will help us improve this in the future. I have a few questions on orchagent restarts in the existing SONiC:
1. You mentioned that the "configs are reprogrammed by managers". Are those the config managers that reads from CONFIG DB and writes to APPL DB? But for L3 forwarding, there is no CONFIG DB. Does anyone read the host routing and program them in APPL DB?
2. When swss container restarts, will syncd container also restarts and wipes the ASIC? If not, how will syncd handle the "duplicate" L3 forwarding requests after orchanger restarts?

Thanks.

Reply:
1. Yes and for BGP, it learns the route. orchagent restart will also restart bgp
2. Yes, syncd container restarts when orchagent restarts

Reply:
Thanks. That's very helpful.

From my understanding, the existing SONiC behavior does not meet our requirement. Especially in syncd restart, which will wipe the ASIC. It will cause packet drop even if we re-program the rules. In our usage, we do not restart syncd and orchagent. If any of them crash, the ASIC will still function (but no new rules can be programmed). The system will be in a "critical state". The controller will be notified and starts to drain and reboot the switch.

I updated the HLD a little bit to clarify that we don't support orchagent restart for P4RT table yet.